colossus/src/main/scala/colossus/controller/Pipe.scala

```diff
@@ -0,0 +1,411 @@
+  case object Ok extends Pushed
+
+  //the item was pushed, but is approaching capacity, proactive backpressure measures should be taken if possible
+  case object Filling extends Pushed
+
+  //the item was successfully pushed but the pipe is not yet ready for more data, trigger is called when it's ready
+  case class Filled(onReady: Signal) extends Pushed
+
+  //the item was not pushed because the pipe is already full, the same trigger
+  //returned when the Filled result was returned is included
+  case class Full(onReady: Signal) extends NotPushed
+
+  //The pipe has been manually closed (without error) and is not accepting any more items
+  case object Closed extends NotPushed
+
+  //The pipe has been terminated or some other error has occurred
+  case class Error(reason: Throwable) extends NotPushed
+}
+
+/**
+ * Wraps 2 sinks and will automatically begin reading from the second only when
+ * the first is empty.  The `None` from the first sink is never exposed.  The
+ * first error reported from either sink is propagated.
+ */
+class DualSource[T](a: Source[T], b: Source[T]) extends Source[T] {
+  private var a_empty = false
+  def pull(cb: Try[Option[T]] => Unit) {
+    if (a_empty) {
+      b.pull(cb)
+    } else {
+      a.pull{
+        case Success(None) => {
+          a_empty = true
+          pull(cb)
+        }
+        case other => cb(other)
+      }
+    }
+  }
+
+  def terminate(reason: Throwable) {
+    a.terminate(reason)
+    b.terminate(reason)
+  }
+
+  override def terminated: Boolean = if (a_empty) b.terminated else a.terminated
+
+  def isClosed = a.isClosed && b.isClosed
+}
+
+
+
+sealed trait PipeException extends Throwable
+class PipeTerminatedException(reason: Throwable) extends Exception("Pipe Terminated", reason) with PipeException
+class PipeStateException(message: String) extends Exception(message) with PipeException
+
+/**
+ * This is a special exception that Input/Output controllers look for when
+ * error handling pipes.  In most cases they will log the error that terminated
+ * the pipe, but for this one exception, the failure will be silent.  This is
+ * basically for situations where a certain amount of data is expected but for
+ * some reason the receiver decides to cancel for some business-logic reason.
+ */
+class PipeCancelledException extends Exception("Pipe Cancelled") with PipeException
+
+class BufferedPipe[T](size: Int, lowWatermarkP: Double = 0.8, highWatermarkP: Double = 0.9) extends Pipe[T, T] {
+
+  sealed trait State
+  sealed trait PushableState extends State
+  case class Full(trigger: Trigger) extends State
+  case class Pulling(callback: Try[Option[T]] => Unit) extends PushableState
+  case object Closed extends State
+  case object Idle extends PushableState
+  case class Dead(reason: Throwable) extends State
+
+  //Full is the default state because we can only push once we've received a callback from pull
+  private var state: State = Idle
```

Comment:
Ah yeah the initial state used to be `Full` because without a buffer the producer could only push an item if the consumer had already asked for one, but that's no longer the case so `Idle` is the correct initial state.